tests/testthat/test-mixedmodelsglmm.R

```diff
@@ -0,0 +1,760 @@
+context("Generalized Linear Mixed Models")
+
+
+### binomial + logit, default, all selected output using LRT method
+{
+  options <- analysisOptions("MixedModelsGLMM")
+  options$.meta <- list(Contrasts = list(list(levels = list(containsColumn = TRUE)), 
```

Comment:
All `.meta` can be removed. jaspTools should do this automatically though o.O How did you generate the tests?

Reply:
I think that I used the new package, but I still have both installed, so, it's possible that I used wrong capitalization when creating them.

Reply:
I deleted them and re-pushed the commit, but many of the unit tests fail, which is very surprising - all of them work perfectly on my PC. I'm getting errors like this:

```
Error in stats::var(validValues) : 
  Calling var(x) on a factor x is defunct.
  Use something like 'all(duplicated(x)[-1L])' to test for a constant vector.
Error in if (any(nearOne)) { : missing value where TRUE/FALSE needed
```

but those are from within the packages and don't make any issues. However, the figures don't match and some tables do not either...

Reply:
Yeah those messages aren't a problem.

Does your analysis
- have seeds set at all required positions (and also have the seed option specified in the options of each the unit test)?
- use any C++ code that performs a random process?

Reply:
> does your analysis have a seed set at all required positions in the R code?

not sure whether that's relevant, the flow should be identical here and on my PC. Also, there is always a `set.seed(1)` call before the `jaspTools::runAnalysis` call.

> does each unit test have the seed option specified (so not set.seed() but the option that comes from qml)?

It wasn't the case, I just added it to the unit tests for LMM but there is no difference.

> Does your analysis use any C++ code that performs a random process and might that ignore the seed?

There are some parts that use C++ for bootstrap, however, there are failing parts that do not bootstrap at all (even though there might be some inner C++ dependency). For example, the following table should be completely seed independent, but still fails on Travis

```
> set.seed(1)
> afex::mixed(formula = as.formula(model_formula$model_formula), 
+             data = dataset,
+             type = options$type,
+             method = options$method, 
+             test_intercept =  FALSE,
+             args_test = list(nsim = options$bootstrap_samples), 
+             check_contrasts = TRUE)
Contrasts set to contr.sum for the following variables: facExperim, facGender
Numerical variables NOT centered on 0: contGamma, contBinom
If in interactions, interpretation of lower order (e.g., main) effects difficult.
Fitting one lmer() model. boundary (singular) fit: see ?isSingular
[DONE]
Calculating p-values. [DONE]
Mixed Model Anova Table (Type 3 tests, S-method)

Model: contNormal ~ contGamma * contBinom * facExperim * facGender + 
Model:     (contGamma + contBinom + facExperim + facGender | facFive)
Data: dataset
                                     Effect       df      F p.value
1                                 contGamma  1, 3.99   0.13    .733
2                                 contBinom 1, 61.47   0.09    .768
3                                facExperim 1, 17.60   2.38    .141
4                                 facGender 1, 25.74   2.40    .134
5                       contGamma:contBinom 1, 75.87   0.34    .563
6                      contGamma:facExperim 1, 56.27 4.62 *    .036
7                      contBinom:facExperim 1, 74.95   2.63    .109
8                       contGamma:facGender 1, 76.98   1.41    .239
9                       contBinom:facGender 1, 77.57   0.77    .382
10                     facExperim:facGender 1, 73.64   1.47    .229
11           contGamma:contBinom:facExperim 1, 78.19 3.30 +    .073
12            contGamma:contBinom:facGender 1, 78.12   1.36    .247
13           contGamma:facExperim:facGender 1, 63.82   0.24    .628
14           contBinom:facExperim:facGender 1, 76.32   1.95    .167
15 contGamma:contBinom:facExperim:facGender 1, 76.13 3.95 +    .050
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘+’ 0.1 ‘ ’ 1
Warning message:
lme4 reported (at least) the following warnings for 'full':
  * boundary (singular) fit: see ?isSingular 
> set.seed(2)
> afex::mixed(formula = as.formula(model_formula$model_formula), 
+             data = dataset,
+             type = options$type,
+             method = options$method, 
+             test_intercept =  FALSE,
+             args_test = list(nsim = options$bootstrap_samples), 
+             check_contrasts = TRUE)
Contrasts set to contr.sum for the following variables: facExperim, facGender
Numerical variables NOT centered on 0: contGamma, contBinom
If in interactions, interpretation of lower order (e.g., main) effects difficult.
Fitting one lmer() model. boundary (singular) fit: see ?isSingular
[DONE]
Calculating p-values. [DONE]
Mixed Model Anova Table (Type 3 tests, S-method)

Model: contNormal ~ contGamma * contBinom * facExperim * facGender + 
Model:     (contGamma + contBinom + facExperim + facGender | facFive)
Data: dataset
                                     Effect       df      F p.value
1                                 contGamma  1, 3.99   0.13    .733
2                                 contBinom 1, 61.47   0.09    .768
3                                facExperim 1, 17.60   2.38    .141
4                                 facGender 1, 25.74   2.40    .134
5                       contGamma:contBinom 1, 75.87   0.34    .563
6                      contGamma:facExperim 1, 56.27 4.62 *    .036
7                      contBinom:facExperim 1, 74.95   2.63    .109
8                       contGamma:facGender 1, 76.98   1.41    .239
9                       contBinom:facGender 1, 77.57   0.77    .382
10                     facExperim:facGender 1, 73.64   1.47    .229
11           contGamma:contBinom:facExperim 1, 78.19 3.30 +    .073
12            contGamma:contBinom:facGender 1, 78.12   1.36    .247
13           contGamma:facExperim:facGender 1, 63.82   0.24    .628
14           contBinom:facExperim:facGender 1, 76.32   1.95    .167
15 contGamma:contBinom:facExperim:facGender 1, 76.13 3.95 +    .050
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘+’ 0.1 ‘ ’ 1
```

Reply:
Interestingly when I run your tests then the results I obtain are identical to what travis reports.
Could you post your session info?

```
> sessionInfo()
R version 3.6.3 (2020-02-29)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS Mojave 10.14.6

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib

Random number generation:
 RNG:     Mersenne-Twister 
 Normal:  Inversion 
 Sample:  Rounding 
 
locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] shiny_1.5.0            jaspMixedModels_0.13.0 gdtools_0.2.2          jaspResults_1.15       jaspBase_0.15         
[6] testthat_2.3.2         jaspTools_1.3.1       

loaded via a namespace (and not attached):
  [1] TH.data_1.0-10          minqa_1.2.4             colorspace_1.4-1        rjson_0.2.20           
  [5] ellipsis_0.3.1          rio_0.5.16              rprojroot_1.3-2         estimability_1.3       
  [9] fs_1.5.0                rstudioapi_0.11         farver_2.0.3            remotes_2.2.0          
 [13] fansi_0.4.1             mvtnorm_1.0-11          codetools_0.2-16        splines_3.6.3          
 [17] pkgload_1.1.0           afex_0.28-0             jsonlite_1.7.0          nloptr_1.2.1           
 [21] vdiffr_0.3.2.2          packrat_0.5.0           pbkrtest_0.4-7          compiler_3.6.3         
 [25] emmeans_1.4.7           backports_1.1.9         fastmap_1.0.1           assertthat_0.2.1       
 [29] Matrix_1.2-17           cli_2.0.2               later_1.1.0.1           htmltools_0.5.0        
 [33] prettyunits_1.1.1       tools_3.6.3             lmerTest_3.1-0          coda_0.19-3            
 [37] gtable_0.3.0            glue_1.4.1              reshape2_1.4.3          Rcpp_1.0.5             
 [41] carData_3.0-2           cellranger_1.1.0        fontquiver_0.2.1        vctrs_0.3.2            
 [45] nlme_3.1-140            stringr_1.4.0           ps_1.3.4                openxlsx_4.1.3         
 [49] lme4_1.1-23             mime_0.9                lifecycle_0.2.0         devtools_2.3.1         
 [53] statmod_1.4.34          MASS_7.3-51.4           zoo_1.8-6               scales_1.1.1           
 [57] promises_1.1.1          jaspGraphs_0.5.1        hms_0.5.2               parallel_3.6.3         
 [61] sandwich_2.5-1          fontLiberation_0.1.0    yaml_2.2.1              curl_4.3               
 [65] memoise_1.1.0           gridExtra_2.3           ggplot2_3.3.2           stringi_1.4.6          
 [69] fontBitstreamVera_0.1.1 desc_1.2.0              boot_1.3-23             pkgbuild_1.1.0         
 [73] zip_2.0.4               freetypeharfbuzz_0.2.5  rlang_0.4.7             pkgconfig_2.0.3        
 [77] systemfonts_0.2.3       lattice_0.20-38         purrr_0.3.4             htmlwidgets_1.5.1      
 [81] labeling_0.3            processx_3.4.3          plyr_1.8.4              magrittr_1.5           
 [85] R6_2.4.1                multcomp_1.4-10         pillar_1.4.6            haven_2.2.0            
 [89] foreign_0.8-71          withr_2.2.0             survival_2.44-1.1       abind_1.4-5            
 [93] ggpol_0.0.6             tibble_3.0.3            crayon_1.3.4            car_3.0-4              
 [97] usethis_1.6.1           grid_3.6.3              readxl_1.3.1            data.table_1.12.8      
[101] callr_3.4.3             forcats_0.4.0           digest_0.6.25           xtable_1.8-4           
[105] httpuv_1.5.4            numDeriv_2016.8-1.1     munsell_0.5.0           sessioninfo_1.1.1 
```

Could you also remake the plots with the latest version of ggplot2/jaspGraphs?

Reply:
omg, the weirdest things are happening.
1) my jastTools just disappeared and I had to reinstall them
2) 3 figures stopped matching, all of the tests were OK, I pushed them to CRAN
3) I see that it fails again on the server, I restart R and now 3 cases just terminate
4) I restart R again, another figure is not matching anymore....

Reply:
Could you verify that:
- jasp-required-files is up-to-date
- the session info after you run the tests each time is the same

Reply:
my jasp-required files seem to be up to date:
![image](https://user-images.githubusercontent.com/38475991/99439914-1d264a00-2916-11eb-9e11-a1c448018b88.png)

the session infos differed when there were different failures, but I have no idea why, since I always cleaned workspace and restarted R. I found that in the more falling session, it loaded a local older version of afex + slightly different version of boot and lmerTest. This is the session that produces only one error and seems to be probably the same as you posted - the key packages look the same.

Also, what's interesting is that the errors on travis are not because of different values inside of the tables but due to missing columns in the tables... I'm completely puzzled by this.

```
R version 3.6.3 (2020-02-29)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19041)

Matrix products: default

Random number generation:
 RNG:     Mersenne-Twister 
 Normal:  Inversion 
 Sample:  Rounding 
 
locale:
[1] LC_COLLATE=English_Europe.1252  LC_CTYPE=English_Europe.1252    LC_MONETARY=English_Europe.1252 LC_NUMERIC=C                    LC_TIME=English_Europe.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] gdtools_0.2.2    jaspResults_1.15 jaspBase_0.15    testthat_3.0.0   jaspTools_1.3.1 

loaded via a namespace (and not attached):
  [1] nlme_3.1-140            pbkrtest_0.4-7          fs_1.5.0                usethis_1.6.3           fontquiver_0.2.1        devtools_2.3.2         
  [7] rprojroot_1.3-2         numDeriv_2016.8-1.1     tools_3.6.3             backports_1.2.0         R6_2.5.0                afex_0.28-0            
 [13] colorspace_1.4-1        withr_2.3.0             tidyselect_0.2.5        gridExtra_2.3           prettyunits_1.1.1       processx_3.4.4         
 [19] ggpol_0.0.6             emmeans_1.4.7           curl_4.2                compiler_3.6.3          cli_2.1.0               vdiffr_0.3.3           
 [25] fontBitstreamVera_0.1.1 desc_1.2.0              sandwich_2.5-1          labeling_0.3            scales_1.0.0            mvtnorm_1.0-11         
 [31] callr_3.5.1             systemfonts_0.3.2       stringr_1.4.0           digest_0.6.27           foreign_0.8-71          minqa_1.2.4            
 [37] rio_0.5.16              pkgconfig_2.0.3         lme4_1.1-23             sessioninfo_1.1.1       rlang_0.4.8             readxl_1.3.1           
 [43] rstudioapi_0.12         zoo_1.8-6               jsonlite_1.6            dplyr_0.8.3             zip_2.0.4               car_3.0-4              
 [49] magrittr_1.5            Matrix_1.2-17           Rcpp_1.0.2              munsell_0.5.0           fansi_0.4.1             abind_1.4-5            
 [55] stringi_1.5.3           multcomp_1.4-10         carData_3.0-2           MASS_7.3-51.4           pkgbuild_1.1.0          plyr_1.8.4             
 [61] grid_3.6.3              parallel_3.6.3          forcats_0.4.0           crayon_1.3.4            lattice_0.20-38         haven_2.1.1            
 [67] splines_3.6.3           hms_0.5.2               zeallot_0.1.0           ps_1.4.0                pillar_1.4.2            boot_1.3-23            
 [73] rjson_0.2.20            estimability_1.3        reshape2_1.4.3          codetools_0.2-16        pkgload_1.1.0           glue_1.4.2             
 [79] fontLiberation_0.1.0    data.table_1.12.6       remotes_2.2.0           vctrs_0.2.0             nloptr_1.2.1            cellranger_1.1.0       
 [85] gtable_0.3.0            purrr_0.3.3             assertthat_0.2.1        ggplot2_3.3.2           jaspMixedModels_0.13.0  openxlsx_4.1.3         
 [91] xtable_1.8-4            jaspGraphs_0.5.1        coda_0.19-3             survival_2.44-1.1       tibble_2.1.3            freetypeharfbuzz_0.2.5 
 [97] lmerTest_3.1-0          memoise_1.1.0           statmod_1.4.34          TH.data_1.0-10          ellipsis_0.3.1    
```

Reply:
> Also, what's interesting is that the errors on travis are not because of different values inside of the tables but due to missing columns in the tables... I'm completely puzzled by this.

No, it is because of changed values. I might have to make that table diff a bit clearer.
Basically you have to compare the new table values to the old table values that appear a few lines below them.
E.g.,
```
Failure (test-mixedmodelslmm.R:237:5): Estimated Trends table results match

New table is not equal to old table:

New table value `-0.254685879186117` (col `lowerCI`, row 1) does not exist in old table

New table value `0.122503695761503` (col `slope`, row 1) does not exist in old table

New table value `0.499693270709123` (col `upperCI`, row 1) does not exist in old table

Old table values `-0.254288816251151`, `0.122785177632287`, `0.499859171515724` do not exist in new table
```

I will "fix" the unit tests so they pass for me and then we can see if we can make them reproducible for you.

Reply:
oh, I see now. Thanks (y)